### ISSUE TITLE
[React 19] Bug : Weird behavior when changing the order of a list

### ISSUE THREAD

**Issue by Armand-Dusart** [facebook/react#31553]
# Explanations :

It seems to have some weird behavior introduced in the React 19 RC update. When we iterate over a list to create components and put a unique id as a key, when sorting certain elements are rerendered. Which was not the case in version 18.

# The effect :

```js
//In each item

useEffect(() => {
  const doc = ref.current;
  if (!doc) {
    return;
  }

  const timeout = setTimeout(() => {
    doc.animate(
      [{ outlineColor: "#d20f39" }, { outlineColor: "transparent" }],
      {
        duration: 300,
        easing: "cubic-bezier(0.4, 0, 0.2, 1)",
        iterations: 2,
      }
    );
  }, 500);

  return () => {
    clearTimeout(timeout);
  };
}, [item.title]);
```

# React 18.3.1 :

https://github.com/user-attachments/assets/bb21513a-5eda-48bd-b306-81215d0a25f2

# React 19.0.0-rc-66855b96-20241106 :

https://github.com/user-attachments/assets/8b229790-56e9-4ace-be38-e4d7656ce61d

# Code
https://github.com/Armand-Dusart/react-19-key-bug

#Meta 
Node : 22.9.0
npm : 10.8.3

**Comment by josephsavona** [comment]
Thanks for posting! Can you share a code sandbox or small GitHub repo that reproduces the error? My guess is that the issue is w how you’re generating key values but without a repro it’s hard to diagnose.

**Comment by Armand-Dusart** [comment]
You can found the code on this github link: https://github.com/Armand-Dusart/react-19-key-bug

´´´js
import React from "react";

import "./App.css";
import List from "./List";
import { faker } from "@faker-js/faker";

const colors = [
  "#f4b8e4",
  "#ca9ee6",
  "#ea999c",
  "#e78284",
  "#99d1db",
  "#a6d189",
  "#e5c890",
  "#85c1dc",
  "#b7bdf8",
  "#94e2d5",
] as const;
// instancing the data 
const data = Array.from({ length: 10 }, (_, i) => ({
  id: i,
  title: faker.lorem.words(),
  description: faker.lorem.paragraph(),
  color: colors[i % colors.length],
}));

function App() {
  return (
    <div className="App">
      <List data={data} />
    </div>
  );
}

export default App;
´´´

If you want a can make a codesandbox later.

**Comment by guillaumebrunerie** [comment]
I can reproduce this issue, and I agree that there is something suspicious going on (or at least something I don’t understand).
Here is a more minimal reproduction (https://codesandbox.io/p/sandbox/v38yzn):
```js
const Item = memo(({ name }) => {
  console.log(`Rendering item ${name}`);
  useEffect(() => {
    console.log(`Running useEffect for item ${name}`);
  }, [name]);
  return <span>{name}</span>;
});

const App = () => {
  const [data, setData] = useState([{ id: "a" }, { id: "b" }]);
  const swap = () => {
    setData((data) => [data[1], data[0]]);
  };
  return (
    <div>
      {data.map((s) => (
        <Item key={s.id} name={s.id} />
      ))}
      <button onClick={swap}>Swap</button>
    </div>
  );
};
```

Click the "Swap" button several times with the console open and notice that:
- the Item components are not rerendered,
- but the effect of one of them **is** run again.

```
Rendering item a
Rendering item a
Rendering item b
Rendering item b
Running useEffect for item a
Running useEffect for item b
Running useEffect for item a
Running useEffect for item b
Running useEffect for item a
Running useEffect for item b
Running useEffect for item a
Running useEffect for item b
Running useEffect for item a
Running useEffect for item b
Running useEffect for item a
Running useEffect for item b
```

It is very unexpected to me that some effects are run even though the corresponding component isn’t rerendered at all. Or is that expected behavior in React 19? It only happens with StrictMode (all the additional console.log are the dimmed version).

**Comment by HarshTripathi1103** [comment]
@guillaumebrunerie This behavior is expected in  Strict Mode. Strict Mode in development intentionally remounts components to help detect potential issues, which causes useEffect to run multiple times even if the component doesn't re-render. This is a development-only feature and won’t happen in production. Removing Strict Mode will stop the repeated useEffect calls.

**Comment by guillaumebrunerie** [comment]
> @guillaumebrunerie This behavior is expected in Strict Mode. Strict Mode in development intentionally remounts components to help detect potential issues, which causes useEffect to run multiple times even if the component doesn't re-render. This is a development-only feature and won’t happen in production. Removing Strict Mode will stop the repeated useEffect calls.

I know about that, but note that here the component is *not* rerendered (or at least the console.log in the body of the component is not run).

**Comment by HarshTripathi1103** [comment]
@guillaumebrunerie I guess it remounts during development and not re-renders to stimulate mount / unmount  cycle to ensure the code handle sideffects properly.This is why useEffect runs again without the component being rerendered. It happens in  development-only and won’t happen in production, which is  weird by the way

**Comment by Armand-Dusart** [comment]
Indeed, without strictMode, the behavior in React 19 disappears. However, note (in StrictMode) that there is a weird change in behavior between version 18 and version 19. Which can mislead the developer in version 19, in my opinion.
